keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,565 @@
+env vars. For example, consider that each node has the following env file:
+
+```
+CONFIG_VAR=HELLO
+CONFIG_VAR_A=WORLD
+...
+```
+
+The user can use the `fileKeyRef` field to select a subset of these env vars
+that it wants to provide to the app. Behind the scene, kubelet will parse the
+env file on the host and instantiate these env vars when creating the main
+container. It does not mount this hostPath volume onto the main container.
+In this case, the Pod's output includes `CONFIG_VAR=HELLO`
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'cp /etc/config/config.env /data/config.env']
+    volumeMounts:
+    - name: config
+      mountPath: /data
+    - name: data
+      mountPath: /etc/config
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: config.env
+          volumeName: config
+          key: CONFIG_VAR
+  restartPolicy: Never
+  volumes:
+  - name: config
+    emptyDir: {}
+  - name: data
+    hostPath: /etc/config
+```
+
+## Design Details
+
+Pods and PodTemplate will include `FileEnvSource` field under [`EnvFromSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvFromSource)
+and `FileKeySelector` field under [`EnvVarSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvVarSource)
+that can be set for an individual container.
+
+### Pod API
+
+`FileEnvSource` field will allow the users to populate environment variables
+from the specified file. The user can specify a file using `VolumeName` and
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+```
+type EnvFromSource struct {
+    ...
+    // The file to select from
+    // +optional
+    FileRef *FileEnvSource `json:"fileRef,omitempty" protobuf:"bytes,3,opt,name=fileRef"`
+    ...
+}
+
+type FileEnvSource struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // Specify whether the file must exist. If the file does not exist,
+    // then the env var is not published.
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,3,opt,name=optional"`
+}
+```
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file.
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
```

Comment:
What if we drop FileRef and allow this to be a list instead? That way the risk of leaking secrets is reduced a bit and the spec is explicit in what it needs.

Reply:
(I would be okay approving with that and then we still have time to debate other possible security concerns before we merge the implementation PR). @thockin @SergeyKanzhelev @tallclair wdyt?

Reply:
Dropped `FileRef` field. Now we will only have `FileKeyRef` field under `env` spec.